### PR TITLE
 #287 Change last update to be saved per provider

### LINKED ISSUE #195: Replace Dota provider plusplus-loop
The `src/providers/provider_dota.ts` file still contains a potentially unnecessary plusplus-loop:
```js
for (let i = 0; i < patchList.length && patchList[i] !== lastPatch; i++) {
  newPatches.push(patchList[i]);
}
```
This should be changed to a different kind of loop.

### LINKED ISSUE #287: Fix Veloren RSS updates not being sent
**Description:**
Currently, the Veloren blog posts are potentially getting lost for the updater. This is because the date registered in the RSS item is not the actual time of the post, but the Monday before. This causes the updater to think they are old posts if any other games were updated in between, discarding them in the update cycle.

**Expected behavior:**
We can either switch the provider source from the RSS feed to the reddit posts or adjust the updater accordingly. Switching the provider source is easier, but less reliable and the blog contents are lost. The updater could be adjusted to save the time of the last update *per game*, instead of having one timestamp for everything. This solution could be implemented together with #239/#252.

### LINKED ISSUE #293: Fix Dota gameplay updates not being sorted correctly
**Description:**
If multiple new gameplay updates are available, the provider takes the oldest instead of the newest updates.

**To Reproduce:**

1. Set `dota.lastVersion` to an old version
2. Set `dota.lastUpdate` to an old date
3. Subscribe to dota and restart
4. The bot will send the oldest gameplay updates instead of the newsest

**Expected behavior:**
The gameplay updates should be sorted correctly, taking the newest updates instead of the oldest.

### PR DESCRIPTION
**Pull request by TimJentzsch**
**Description**:
- Changes last update to be saved per provider instead of per updater. This closes #287.
- Fixes Dota gameplay patches not being sorted properly. This closes #293.
- Removes Dota provider plusplus-loop. This closes #195.

**Checklist**:

- [ ] Tested the bot on both clients with a few commands
- [ ] Updated the [`README`](README.md) if necessary
- [ ] Updated the version string in the [`package.json`](package.json) if necessary

### CODE REVIEW

**Review by zachkont — APPROVED** [review]

File: src/providers/dota_provider.ts [hunk 1]
```diff
@@ -21,68 +19,48 @@ export default class DotaProvider extends Provider {
...
+        // Sort ascending (latest patches are last)
+        .sort((a, b) => {
+          if (a < b) {
+            return -1;
+          }
+          if (a > b) {
+            return 1;
+          }
+          return 0;
+        })
```
Comment (zachkont) [inline src/providers/dota_provider.ts:42]: This doesn't work perfectly, for example "9.12c" is larger than "10.1a"

File: src/updater.ts [hunk 2]
```diff
@@ -92,81 +91,66 @@ export default class Updater {
...
-      const pollTime = endPollTime - startTime;
-      this.logger.info(
-        `Found ${notifications.length} posts in ${pollTime} ms. Notifying channels...`,
-      );
-
-      // Update time
-      this.saveDate(notifications[notifications.length - 1].timestamp);
+      // Get the updates for the current game
+      await this.updateGame(game);
+    }, Promise.resolve());
```
Comment (zachkont) [inline src/updater.ts:105]: I don't find this more readable than a simple for loop

Comment (TimJentzsch) [reply src/updater.ts]: Fixed

**Comment by zachkont** [comment]
test comment from vscode plugin

File: src/providers/provider.ts [hunk 3]
```diff
@@ -15,5 +19,74 @@
     this.logger = new Logger(this.label);
   }
 
-  public abstract async getNotifications(date?: Date, limit?: number): Promise<Notification[]>;
+  public abstract async getNotifications(updater: Updater, limit?: number): Promise<Notification[]>;
```
Comment (zachkont) [inline src/providers/provider.ts:22]: test comment

File: src/notifications/version.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+export default class Version implements Comparable<Version> {
+  /**
+   * Creates a new Version.
+   * @param version The gameplay version of the update.
+   */
+  constructor(public version: string) {}
+
+  public compareTo(other: Version): 0 | 1 | -1 {
```
Comment (zachkont) [inline src/notifications/version.ts:37]:
```suggestion
  public compareTo(other: Version): -1 | 0 | 1 {
```

File: src/updater.ts [hunk 5]
```diff
@@ -90,19 +90,18 @@ export default class Updater {
...
+      // Delay in between game
+      if (index < games.length - 1) {
+        // eslint-disable-next-line no-await-in-loop
         await sleep(this.gameIntervalMs);
       }
-
-      // Get the updates for the current game
-      await this.updateGame(game);
-    }, Promise.resolve());
+    }
```
Comment (zachkont) [inline src/updater.ts]: Super nitpick but I would do it like this:

```ts
 if (index > 0) {
  await sleep(this.gameIntervalMs);
}

await this.updateGame(game);
```

Comment (TimJentzsch) [reply src/updater.ts]: Fixed

File: src/providers/dota_provider.ts (hunk 1 above)
Comment (TimJentzsch) [reply src/providers/dota_provider.ts]: Fixed

**Pull request merged**